drv/lpc55-spi-server/src/main.rs

```diff
@@ -8,24 +8,329 @@
 #![no_std]
 #![no_main]
 
+use drv_lpc55_gpio_api::{Direction, Pin, Value};
 use drv_lpc55_spi as spi_core;
 use drv_lpc55_syscon_api::{Peripheral, Syscon};
+use drv_spi_msg::*;
 use lpc55_pac as device;
 use ringbuf::*;
+
+use crc::{Crc, CRC_32_CKSUM};
+use lpc55_romapi::bootrom;
 use userlib::*;
 
 task_slot!(SYSCON, syscon_driver);
 task_slot!(GPIO, gpio_driver);
 
+// See drv/spi-msg/README.md
+// Messages are received from the Service Processor (SP).
+//
+// Messages are framed at the physical layer by the SP asserting the
+// `Chip Select` (negated) or `CSn` signal.
+//
+// Although SPI supports full duplex communications, the messages carried
+// between the SP and the RoT may take significant time for the RoT to process.
+// Therefore, some form of flow control is needed.
+// The RoT has control of the ROT_IRQ line which is used to indicate that the
+// RoT has a message ready for the SP to read.
+//
+// TODO: SP should be able to send a message to resync the RoT communications.
+// To support that semantic, the RoT will always
+// process incoming bytes even if it is transmitting a previously queued
+// message.
+// TODO: properly test that case.
+//
+// Messages for protocol 1 are composed of a four-byte header:
+//   - a protocol version (0x01)
+//   - the lease significant byte of the payload length
+//   - the most significant byte of the payload length
+//   - a message type
+//   - the payload, not to exceed a maximum length
+//
+// If the payload length exceeds the maximum size or not all bytes are received
+// before CSn is de-asserted, the message is malformed and an error message
+// will be queued if Tx is idle.
+//
+// RoT Tx messages will not be written to the Tx FIFO until CSn is de-asserted.
+// ROT_IRQ will not be asserted until data has been written to the Tx FIFO.
+// This allows the RoT to take an arbitrary amount of time to prepare a
+// response message and allows the RoT to make a best effort to meet
+// the real time requirements of feeding the Tx FIFO when the SP clocks out
+// the response message.
+//
+// ROT_IRQ is expected to be an edge triggered interrupt on the SP.
+// ROT_IRQ is de-asserted only after all Tx bytes have drained.
+// That gives a bit more information when looking at logic analyzer traces.
+
+#[derive(Copy, Clone, PartialEq)]
+#[repr(u8)]
+enum RxState {
+    Idle = 0,       // Idle, waiting for message
+    Header = 1,     // Expecting message length LSB, MSB, and MsgType
+    Payload = 2,    // receiving message payload
+    Dispatch = 3,   // Receive complete, message processing is pending.
+    Responding = 4, // Tx frame is queued. Terminal state until next SOT.
+    Invalid = 5,    // bad message or normally ignoring incoming data until SOT.
+    Error = 0xff,   // Errors, Terminal state until next SOT.
+}
+
+impl From<u8> for RxState {
```

Comment:
Are we encoding `RxState` and `TxState` as a `u8` anywhere? If not, I'd suggest converting them to pure enums – remove the explicit integer values, `#[repr(u8)]`, and this `From` implementation, since we aren't using any of that.

Reply:
u8 encoding is not necessary. From for Rx/TxState also not necessary.